File: lib-core/src/main/java/bas/lib/core/converter/Converters.java
package bas.lib.core.converter;

import bas.lib.core.converter.fastjson.FastJsonConverter;
import bas.lib.core.converter.gson.GsonConverter;
import bas.lib.core.converter.jackson.JacksonConverter;
import bas.lib.core.lang.ClassUtils;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Created by Lucio on 2021/7/21.
 */
public class Converters {

    private static JsonConverter mJsonConverter;

    private Converters() {
    }

    static {
        try {
            //初始化默认JsonConverter，优先使用Jackson(Kotlin版本)
            if (ClassUtils.isClassExists("com.fasterxml.jackson.module.kotlin.KotlinModule")) {
                System.out.println("Converters：使用JacksonKotlinConverter");
                mJsonConverter = new bas.lib.core.converter.jackson.JacksonConverter();
            } else if (ClassUtils.isClassExists("com.fasterxml.jackson.databind.ObjectMapper")) {
                System.out.println("Converters：使用JacksonConverter");
                mJsonConverter = new JacksonConverter();
            } else if (ClassUtils.isClassExists("com.google.gson.Gson")) {
                System.out.println("Converters：使用GsonConverter");
                mJsonConverter = new GsonConverter();
            } else if (ClassUtils.isClassExists("com.alibaba.fastjson.JSON")) {
                System.out.println("Converters：使用FastJsonConverter");
                mJsonConverter = new FastJsonConverter();
            } else {
                System.out.println("Converters：null");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void setJsonConverter(@NotNull JsonConverter converter) {
        mJsonConverter = converter;
    }

    public static JsonConverter getJsonConverter() {
        return mJsonConverter;
    }

    @Nullable
    public static <T> T toObject(@Nullable String json, Class<T> clazz) {
        return mJsonConverter.toObject(json, clazz);
    }

    @Nullable
    public static <T> List<T> toObjectList(@Nullable String json, Class<T> clazz) {
        return mJsonConverter.toObjectList(json, clazz);
    }

    @Nullable
    public static String toJson(@Nullable Object obj) {
        return mJsonConverter.toJson(obj);
    }

}

File: bas-core-ktx/src/main/java/com/bas/core/converter/Converters.kt
package com.bas.core.converter

import com.bas.core.converter.gson.GsonConverter
import com.bas.core.converter.jackson.JacksonConverter
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import java.lang.reflect.Type

/**
 * Created by Lucio on 2021/7/22.
 */

inline fun Any?.toJson(): String? {
    return Converters.toJson(this)
}

inline fun <reified T> String?.toObject(): T? {
    return Converters.toObject(this, T::class.java)
}

inline fun <reified T> String?.toObjectList(): List<T>? {
    if (this.isNullOrEmpty())
        return null
    val convert = Converters.getJsonConverter()
    return if (convert is GsonConverter && convert.isInlineOptimization) {
        //gson不支持泛型方法，所以必须在内联方法中实现；
        val type: Type = object : TypeToken<List<T>>() {}.type
        convert.gson.fromJson(this, type)
    } else {
        convert.toObjectList(this, T::class.java)
    }
}

/**
 * 使用UTC时间格式
 */
inline fun ObjectMapper.applyUTCDateTimeFormat(): ObjectMapper {
    return JacksonConverter.applyUTCDateFormat(this)
}

/**
 * 使用UTC时间格式
 */
inline fun GsonBuilder.applyUTCDateTimeFormat(): GsonBuilder {
    return GsonConverter.applyUTCDateFormat(this)
}

/**
 * 启用[com.bas.core.converter.gson.LocalJsonField]过滤策略
 */
inline fun GsonBuilder.applyLocalFieldStrategy(): GsonBuilder {
    return GsonConverter.applyLocalFieldStrategy(this)
}

